src/firewall/core/rich.py

```diff
@@ -458,11 +460,19 @@ def check(self, family=None):
             # iptables (v1.4.21) doesn't accept 1/d
             raise FirewallError(errors.INVALID_LIMIT, "%s too slow" % self.value)
 
-    def __str__(self):
-        return 'limit value="%s"' % (self.value)
+        return rate, mult, duration
+
+    def value_normalized(self, check=True):
+        try:
+            rate, mult, duration = self.value_parse()
+        except FirewallError:
+            if check:
+                raise
+            return self.value
+        return f"{rate}/{duration}"
```

Comment:
Why don't we just always make `self.value` normalized? I don't follow why we would not want that. I don't see the value in `value_normalized()` and `value_parse()` when we could just always guarantee that `foo.value` is normalized. If you need the rate the set `self.rate`.

Reply:
So far, the `_Rich_Entry` objects are plain, stateless data holders. They get initialized once, and are not mutated afterwards.

While `Rich_Rule._import_from_string()` necessarily does some parsing and validation, once it got the basic tokens it just creates the `Rich_Rule` and `_Rich_Entry` objects. The entries in there are mostly unvalidated yet (aside what was necessary while parsing the string). Also, the GUI constructs entries, but doesn't yet perform a full validation (that happens during `check()`).

a) It feels wrong, that the constructor of a `_Rich_Entry` would already perform validation. That is, `Rich_Limit(value='foo')` should not raise an exception, that happens later during `check()`. Also, if you set `Rich_Limit(value=' 1 /m')` it feels wrong to mutate that to `entry.value='1/m'` before calling `check()`.

b) these `_Rich_Entry` instances should be immutable after being created. Immutable things are great, it means that entries don't have an unchecked and checked state. Normalizing `self.value` e.g. during `Rich_Limit.check()` seems very wrong, because then the instance mutates.

So the normalization shouldn't happen in `__init__()` and it shouln't happen in `check()`. There is no other good place. Note that the amount of code is still the same. It's not that normalizing `self.value` would safe any code. It might safe multiple parsing, if you are concerned about the overhead of that...

Reply:
In this PR you argue we should not be normalizing, but you also have PR #1290 that is all about normalizing configuration.

I'm pushing back on this PR. Just normalize when the object is created. I have not seen any reasoning or justification about the benefits of not normalizing when the object is created.

Even the standard library throws errors at object creation.

```
>>> import ipaddress.IPv4Address as IPv4Address
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'ipaddress.IPv4Address'; 'ipaddress' is not a package
>>> from ipaddress import IPv4Address
>>> IPv4Address('not an address')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/egarver/pkg-2023Q3/lib/python3.11/ipaddress.py", line 1315, in __init__
    self._ip = self._ip_int_from_string(addr_str)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/egarver/pkg-2023Q3/lib/python3.11/ipaddress.py", line 1202, in _ip_int_from_string
    raise AddressValueError("Expected 4 octets in %r" % ip_str)
ipaddress.AddressValueError: Expected 4 octets in 'not an address'
>>> IPv4Address('10.10.10.10')
IPv4Address('10.10.10.10')
```

Reply:
> In this PR you argue we should not be normalizing, but you also have PR #1290 that is all about normalizing configuration.

This branch also did normalization. Except, it didn't do it in `Rich_Rule.__init__()`.

Changed now. The constructor normalizes the values.

>  Even the standard library throws errors at object creation.

I didn't say that `__init__()` never fails. However, in this case it seems not best, because there is a `check()` function that fails at a more appropriate moment. That was the case already before the patch, and it still is.